### PR TITLE
Introduce package generators

### PR DESCRIPTION
**Pull request by arthur791004**
#### Proposed Changes

I'm not sure how we add the new sub-package. For me, I just copied and pasted the existing package, and renamed it to the new package name... 🙃

As a result, I made a generator to create it. Any thoughts?

* Define the templates of sub-packages
* Create a CLI command to generate sub-packages inside the wp-calypso monorepo

#### Testing Instructions

* Run the following commands and see whether the generated package looks as expected
  ```bin
  $ yarn
  $ yarn run generate
  ```

#### Pre-merge Checklist

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

File: packages/generators/.eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+	rules: {
+		'import/no-nodejs-modules': 0,
```
Comment (noahtallen) [inline packages/generators/.eslintrc.cjs]: I think we have a default eslint config that can be used for Node scripts. It's used here, for example:

https://github.com/Automattic/wp-calypso/blob/c2a3f57266c1bb6123cd9b474491c31f42fca6a9/packages/effective-module-tree/.eslintrc.js#L1-L4

File: package.json [hunk 2]
```diff
@@ -94,6 +94,7 @@
 		"docker-jetpack-cloud": "docker run -it --env CALYPSO_ENV=jetpack-cloud-production --name wp-calypso --rm -p 80:3000 wp-calypso",
 		"eslint-branch": "node bin/eslint-branch.js",
 		"feature-search": "node bin/feature-search.js",
+		"generate": "wp-generate",
```
Comment (noahtallen) [inline package.json]: I feel like `wp-*` might not be a good name for the binary, since the packages aren't really related to WordPress

File: packages/generators/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Generator
+
+Generate a sub-package under wp-calypso
+
+## Usage
+
+```
+npm run generate
```
Comment (noahtallen) [inline packages/generators/README.md]: Let's add `yarn` as an example

File: packages/generators/generators/component.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { actions, prompts } from './constants.js';
+
+const componentGenerator = {
+	description: 'Add a component package',
+	prompts,
+	actions: () => [
+		...actions,
+		{
+			type: 'add',
+			path: '../{{kebabCase name}}/src/index.tsx',
```
Comment (noahtallen) [inline packages/generators/generators/component.js]: Thanks for defaulting to Typescript here!

File: packages/generators/package.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		"generate": "wp-generate"
+	},
+	"dependencies": {
+		"commander": "^5.1.0",
+		"debug": "^4.3.3"
+	},
+	"devDependencies": {
+		"@automattic/calypso-eslint-overrides": "workspace:^",
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"plop": "^3.1.1"
```
Comment (noahtallen) [inline packages/generators/package.json]: `plop` should be a normal dependency, since the library relies on it to work. Dev mode dependencies should just be deps which are not needed after a package is built and published to npm, for example.

File: packages/generators/package.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	},
+	"bugs": "https://github.com/Automattic/wp-calypso/issues",
+	"homepage": "https://github.com/Automattic/wp-calypso/tree/HEAD/packages/generators#readme",
+	"type": "module",
+	"engines": {
+		"node": "^16.17.0",
+		"yarn": "^3.1.1"
+	},
+	"scripts": {
+		"generate": "wp-generate"
```
Comment (noahtallen) [inline packages/generators/package.json]: I might rename `wp-generate` to just `generate-package`? Or `generate-calypso-package` to be more verbose?

Comment (noahtallen) [reply packages/generators/package.json]: and the same on line 14.

File: packages/generators/package.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	},
+	"dependencies": {
+		"commander": "^5.1.0",
+		"debug": "^4.3.3"
```
Comment (noahtallen) [inline packages/generators/package.json]: Do we use commander or debug anywhere in this package? I didn't notice them, but I might have missed it!

File: packages/generators/templates/shared/.eslintrc.js.hbs [hunk 8]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
Comment (noahtallen) [inline packages/generators/templates/shared/.eslintrc.js.hbs]: I think we should take a second look at this file. If this is important common configuration for all packages, we should add it to the root eslintrc file.

**Review by noahtallen — COMMENTED** [review]
Cool! Sorry for taking a while to look at this; it fell through the cracks in my notifications.

Left several small comments, but this is a cool direction and I like the idea. There are some other types of projects we could add in the future, too, like "calypso apps".

File: packages/generators/templates/shared/.eslintrc.js.hbs (hunk 8 above)
Comment (arthur791004) [reply packages/generators/templates/shared/.eslintrc.js.hbs]: I saw lots of packages define this rule, so I think we can add it to the root .eslintrc file and keep this file empty. What do you think?

**Comment by arthur791004** [comment]
@noahtallen Thanks for the review! I have addressed your feedback, and the only thing we need to discuss is https://github.com/Automattic/wp-calypso/pull/69883#discussion_r1053123926 🙂

File: packages/generators/templates/shared/.eslintrc.js.hbs (hunk 8 above)
Comment (noahtallen) [reply packages/generators/templates/shared/.eslintrc.js.hbs]: Hm. On second look, I wonder if we should just remove it and allow a package to add it if needed. Looking at the rule, it kind of seems like it should be very rarely used: https://github.com/jest-community/eslint-plugin-jest/blob/main/docs/rules/no-mocks-import.md, in which case we don't need to add it for every new package

Comment (noahtallen) [reply packages/generators/templates/shared/.eslintrc.js.hbs]: If you agree, I'd recommend removing this file

**Comment by noahtallen** [comment]
You might need to run `yarn dedupe` locally to fix the unit tests

File: packages/generators/templates/shared/package.json.hbs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	},
+	"peerDependencies": {
+		"@wordpress/data": "^6.7.0",
+		"@wordpress/element": "^4.5.0",
+		"@wordpress/i18n": "^4.7.0",
+		"debug": "^4.3.3",
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2",
+		"reakit-utils": "^0.15.1",
+		"redux": "^4.1.2"
```
Comment (noahtallen) [inline packages/generators/templates/shared/package.json.hbs]: I think we might want to modify these somewhat. While I understand the idea, I think we should just add the bare minimum for the package to work out of the box. I don't think it's likely for every new package to need every one of these dependencies.

Additionally, some of these kind of conflict. For example, you don't need react as both a peer dependency and a dev dependency, iirc. And react should probably be a proper dependency rather than a dev dependency.

What I'd suggest is just the minimum packages needed for the build to work, and then authors can add anything else they might need

File: packages/generators/templates/shared/.eslintrc.js.hbs (hunk 8 above)
Comment (arthur791004) [reply packages/generators/templates/shared/.eslintrc.js.hbs]: Removed 👍

File: packages/generators/templates/shared/package.json.hbs (hunk 9 above)
Comment (arthur791004) [reply packages/generators/templates/shared/package.json.hbs]: Sounds good! I've modified the deps and created different `package.json` files for the library and component. Now I only include the react, react-dom deps to the component package. Does it make sense?

File: packages/generators/templates/component/package.json.hbs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+		"react-dom": "^17.0.2",
+		"tslib": "^2.3.0"
+	},
+	"devDependencies": {
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^4.7.4"
+	},
+	"peerDependencies": {
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2"
```
Comment (noahtallen) [inline packages/generators/templates/component/package.json.hbs]: We don't need to set these as peer dependencies if they're defined as normal dependencies

**Review by noahtallen — COMMENTED** [review]
This is looking pretty good! I'll reach out to the rest of the team to get some more eyes on this

**Comment by noahtallen** [comment]
I also found this, which we might want to incorporate: https://github.com/Automattic/wp-calypso/blob/1276e7ecb60e62b3784098c411718bafae2e6b7c/docs/component-readme-template.md

**Comment by noahtallen** [comment]
Finally, to fix the yarn.lock conflict:
1. checkout this branch
2. `git rebase trunk`
3. It'll fail with a message about conflicts. Run `yarn` at this point to automatically fix it.
4. `git add yarn.lock && git rebase --continue`

**Comment by arthur791004** [comment]
Rebased onto trunk, thanks!

File: package.json [hunk 11]
```diff
@@ -213,6 +214,7 @@
 		"@automattic/calypso-eslint-overrides": "workspace:^",
 		"@automattic/calypso-jest": "workspace:^",
 		"@automattic/eslint-plugin-json": "workspace:^",
+		"@automattic/generators": "workspace:^",
```
Comment (tyxla) [inline package.json]: Noting that as an external user / spectator / consumer, the name `@automattic/generators` adds a vagueness, since it's not immediately clear what the package will generate. Should it be `@automattic/package-generators`? One more thing - why the plural? I'd consider the idea being having one generator, with multiple routes for generating packages, but still, one generator. WDYT?

File: packages/generators/README.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+# Generator
+
+Generate a sub-package under wp-calypso
```
Comment (tyxla) [inline packages/generators/README.md]: This is a great place to mention the Calypso monorepo and maybe link to its docs: 

https://github.com/Automattic/wp-calypso/blob/trunk/docs/monorepo.md#L1

File: packages/generators/generators/constants.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
+import packageExists from '../utils/packageExists.js';
+
+export const prompts = [
+	{
+		type: 'input',
+		name: 'name',
+		message: 'What should it be called?',
```
Comment (tyxla) [inline packages/generators/generators/constants.js]: I'd rephrase this to something like:

```suggestion
		message: 'What should the package be called?',
```

There's a substantial improvement when we use concrete words instead of "it", "this" or "that" when explaining / describing things in software development. Tanya Reilly suggests this also in the book [The Staff Engineer's Path](https://www.oreilly.com/library/view/the-staff-engineers/9781098118723/).

File: packages/generators/generators/component.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import { actions, prompts } from './constants.js';
+
+const componentGenerator = {
+	description: 'Add a component package',
```
Comment (tyxla) [inline packages/generators/generators/component.js]: What is a component package? I thought components in the monorepo are supposed to live in `@automattic/components`?

File: packages/generators/README.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+
+## Usage
+
+```
+yarn run generate
```
Comment (tyxla) [inline packages/generators/README.md]: Maybe it's worth covering the different types of templates and providing some examples and usage. Using existing packages will suffice IMHO.

File: packages/generators/package.json [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+	"engines": {
+		"node": "^16.17.0",
+		"yarn": "^3.1.1"
+	},
+	"scripts": {
+		"generate": "generate-calypso-package"
+	},
+	"dependencies": {
+		"plop": "^3.1.1"
+	},
```
Comment (tyxla) [inline packages/generators/package.json]: Is this package intended to be published? If not, maybe we should make it `"private": true`.

File: packages/generators/templates/component/package.json.hbs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+	"dependencies": {
+		"react": "^17.0.2",
+		"react-dom": "^17.0.2",
+		"tslib": "^2.3.0"
+	},
+	"devDependencies": {
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^4.7.4"
+	},
+	"private": true
```
Comment (tyxla) [inline packages/generators/templates/component/package.json.hbs]: We don't always intend to make new packages private, do we? In that case, it might be useful to be able to configure that at package creation.

File: packages/generators/templates/library/package.json.hbs [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+		"watch": "tsc --build ./tsconfig.json --watch"
+	},
+	"dependencies": {
+		"tslib": "^2.3.0"
+	},
+	"devDependencies": {
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^4.7.4"
+	},
+	"private": true
```
Comment (tyxla) [inline packages/generators/templates/library/package.json.hbs]: Same as above.

File: packages/generators/templates/shared/jest.config.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+	preset: '../../test/packages/jest-preset.js',
+	testEnvironment: 'jsdom',
+	testMatch: [ '<rootDir>/**/__tests__/**/*.[jt]s?(x)', '!**/.eslintrc.*' ],
+	setupFilesAfterEnv: [
+		'@testing-library/jest-dom/extend-expect',
+		'@automattic/calypso-build/jest/mocks/match-media',
```
Comment (tyxla) [inline packages/generators/templates/shared/jest.config.js]: Do we really need this one? Seems like it's pretty specific and not necessary?

File: packages/generators/utils/packageExists.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+import path from 'path';
+import { fileURLToPath } from 'url';
+
+const __dirname = path.dirname( fileURLToPath( import.meta.url ) );
+
+const packagePath = path.join( __dirname, '../../' );
+
+const packageNames = fs.readdirSync( packagePath );
+
+const packageExists = ( packageName ) => packageNames.includes( packageName );
```
Comment (tyxla) [inline packages/generators/utils/packageExists.js]: I wonder if this should be a bit more clever, though. Yes, many of the packages exist in the monorepo, and the package name matches the package dir name, however, that's not always the case. There are a bunch of `@automattic` packages that are outside of the repo and within the `Automattic` GitHub organization. Should we make and maintain a list of them and ensure that if someone wants to create a package, it doesn't exist already in the GitHub org (therefore in npm)?

**Review by tyxla — COMMENTED** [review]
Thank you for this great work. I'm sure this will save a lot of time in the future 🚀 

I've left some thoughts and questions, let me know what you think.

I think in any case, this should be in close to a mergeable state and we should be able iterate on it in subsequent PRs, since it doesn't affect production at all. Let's just make sure we don't break anything (like, unit tests currently seem broken for some reason)

File: package.json (hunk 11 above)
Comment (arthur791004) [reply package.json]: Sounds good! I named the package to `@automattic/calypso-package-generator`. WDYT?

File: packages/generators/README.md (hunk 12 above)
Comment (arthur791004) [reply packages/generators/README.md]: Added reference you mentioned to the README.md

File: packages/generators/generators/component.js (hunk 14 above)
Comment (arthur791004) [reply packages/generators/generators/component.js]: I've updated copy to `Add a new library of React components designed for use in Automattic products`, WDYT? We can iterate this in the follow-up PR

File: packages/generators/README.md (hunk 15 above)
Comment (arthur791004) [reply packages/generators/README.md]: Updated!

File: packages/generators/package.json (hunk 16 above)
Comment (arthur791004) [reply packages/generators/package.json]: Okay, I've made it private

File: packages/generators/templates/component/package.json.hbs (hunk 17 above)
Comment (arthur791004) [reply packages/generators/templates/component/package.json.hbs]: I added a new question to ask people "Do you want to publish this package?", and set the value according to the result

File: packages/generators/templates/shared/jest.config.js (hunk 19 above)
Comment (arthur791004) [reply packages/generators/templates/shared/jest.config.js]: Removed!

File: packages/generators/utils/packageExists.js (hunk 20 above)
Comment (arthur791004) [reply packages/generators/utils/packageExists.js]: Sounds good! We can add a list of `@automattic` packages outside this repo in the follow-up PR 👍

File: package.json (hunk 11 above)
Comment (tyxla) [reply package.json]: Thanks! My opinion is that the `calypso-` part could feel redundant, but I can see why it can be useful, so leaving this to your judgment.

File: packages/generators/generators/component.js (hunk 14 above)
Comment (tyxla) [reply packages/generators/generators/component.js]: Sounds better, although I'm still not convinced there's a use case for a similar instance of a package, given that we already have `@automattic/components`.

File: packages/generators/utils/packageExists.js (hunk 20 above)
Comment (tyxla) [reply packages/generators/utils/packageExists.js]: Follow-up for that sounds good 👍 Thanks!

File: packages/calypso-package-generator/README.md [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+Use the following command to run the calypso-package-generator.
+
+```
+yarn run generate
+```
+
+Now we have 2 generators as below, and the calypso-package-generator will ask you to choose.
+
+* component: Add a new library of React components designed for use in Automattic products.
+* library: Add a new library of js designed for use in Automattic products.
```
Comment (tyxla) [inline packages/calypso-package-generator/README.md]: I think the linter will be unhappy about using `*`:

```suggestion
- component: Add a new library of React components designed for use in Automattic products.
- library: Add a new library of js designed for use in Automattic products.
```

File: packages/generators/generators/component.js (hunk 14 above)
Comment (noahtallen) [reply packages/generators/generators/component.js]: I think there are examples of larger features (which are react components) which exist outside of @automattic/components. For example, `@automattic/page-pattern-modal`

Comment (tyxla) [reply packages/generators/generators/component.js]: I see, thanks for clarifying. I guess it's a matter of perspective, since everything in React is either a hook or a component. Let's leave this as-is 👍

**Comment by arthur791004** [comment]
@noahtallen @tyxla Thanks for your reviews! I've rebased this PR onto the latest trunk and fixed the lint. Hence, I think this PR is ready to merge and we can address https://github.com/Automattic/wp-calypso/pull/69883#discussion_r1073494306 in the follow-up PRs 🙂

**Review by tyxla — APPROVED** [review]
Great work @arthur791004, thank you! 💯 

Let's ship and iterate in subsequent PRs 🚀
